pyface/data_view/i_data_view_widget.py

```diff
@@ -8,52 +8,283 @@
+
+    #: Whether the selection is currently being updated.
+    _selection_updating_flag = Bool
+
+    #: The selected indices in the view.
+    _selection = List(Tuple, comparison_mode=ComparisonMode.identity)
+
+    # ------------------------------------------------------------------------
+    # MDataViewWidget Interface
+    # ------------------------------------------------------------------------
+
+    def _header_visible_updated(self, event):
+        """ Observer for header_visible trait. """
+        if self.control is not None:
+            self._set_control_header_visible(event.new)
+
+    def _get_control_header_visible(self):
+        """ Toolkit specific method to get the visibility of the header. """
+        raise NotImplementedError()
+
+    def _set_control_header_visible(self, control_header_visible):
+        """ Toolkit specific method to set the visibility of the header. """
+        raise NotImplementedError()
+
+    def _selection_type_updated(self, event):
+        """ Observer for selection_type trait. """
+        if self.control is not None:
+            self._set_control_selection_type(event.new)
+
+    def _get_control_selection_type(self):
+        """ Toolkit specific method to get the selection type. """
+        raise NotImplementedError()
+
+    def _set_control_selection_type(self, selection_type):
+        """ Toolkit specific method to change the selection type. """
+        raise NotImplementedError()
+
+    def _selection_mode_updated(self, event):
+        """ Observer for selection_mode trait. """
+        if self.control is not None:
+            self._set_control_selection_mode(event.new)
+
+    def _get_control_selection_mode(self):
+        """ Toolkit specific method to get the selection mode. """
+        raise NotImplementedError()
+
+    def _set_control_selection_mode(self, selection_mode):
+        """ Toolkit specific method to change the selection mode. """
+        raise NotImplementedError()
+
+    def _selection_updated(self, event):
+        """ Observer for selection trait. """
+        if self.control is not None and not self._selection_updating_flag:
+            with self._selection_updating():
+                self._set_control_selection(self.selection)
+
+    def _get_control_selection(self):
+        """ Toolkit specific method to get the selection. """
+        raise NotImplementedError()
+
+    def _set_control_selection(self, selection):
+        """ Toolkit specific method to change the selection. """
+        raise NotImplementedError()
+
+    def _observe_control_selection(self, remove=False):
+        """ Toolkit specific method to watch for changes in the selection. """
+        raise NotImplementedError()
+
+    def _update_selection(self, *args, **kwargs):
+        if not self._selection_updating_flag:
+            with self._selection_updating():
+                self._selection = self._get_control_selection()
```

Comment:
It is not pure query: we are setting the `_selection` trait which then trigger `_selection_updated` which needs to have the selection updating flag set to avoid excess work.